FeatureEngine.py

```diff
@@ -42,37 +44,243 @@ def map_labels_to_invoice_OCR(self, data_path: str, verbose: bool = False):
 
     @classmethod
     def create_features(self, token: Token, invoicePage: InvoicePage):
+
+        ################################### HELPER FUNCTIONS FOR TOKEN GENERATION #################################################
+        # calculates and returns min dist from token 1 to token 2 in pixels
+        def calc_min_dist(t1, t2):
+            # get bounding outer rectangle
+            outer_rect_left = min(t1.coordinates["x"], t2.coordinates["x"])
+            outer_rect_top = min(t1.coordinates["y"], t2.coordinates["y"])
+            outer_rect_bottom = max(
+                (t1.coordinates["y"] + t1.coordinates["height"]),
+                (t2.coordinates["y"] + t2.coordinates["height"]),
+            )
+            outer_rect_right = max(
+                (t1.coordinates["x"] + t1.coordinates["width"]),
+                (t2.coordinates["x"] + t2.coordinates["width"]),
+            )
+            outer_rect_width = outer_rect_right - outer_rect_left
+            outer_rect_heigth = outer_rect_bottom - outer_rect_top
+            inner_rect_width = max(
+                0,
+                outer_rect_width - (t1.coordinates["width"] + t2.coordinates["width"]),
+            )
+            inner_rect_height = max(
+                0,
+                outer_rect_heigth
+                - (t1.coordinates["height"] + t2.coordinates["height"]),
+            )
+            pixel_dist = math.sqrt(inner_rect_width ** 2 + inner_rect_height ** 2)
+            return pixel_dist
+
+        # checks if two tokens are aligned vertically within a margin of error (checks midpoint, left boundary, right boundary)
+        def is_vert_aligned(t1, t2, moe):
+            if abs(t1.coordinates["x"] - t2.coordinates["x"]) < moe:
+                return True
+            if (
+                abs(
+                    (t1.coordinates["x"] + t1.coordinates["width"])
+                    - (t2.coordinates["x"] + t2.coordinates["width"])
+                )
+                < moe
+            ):
+                return True
+            t1_midpt_x = t1.coordinates["x"] + (t1.coordinates["width"] / 2)
+            t2_midpt_x = t2.coordinates["x"] + (t2.coordinates["width"] / 2)
+            if abs(t1_midpt_x - t2_midpt_x) < moe:
+                return True
+            return False
+
+        # checks if two tokens are aligned vertically within a margin of error (checks midpoint, top boundary, bottom boundary)
```

Comment:
Typo for comment here -> should be aligned "horizontally"

Reply:
oh yes